### PR TITLE
Geofencing in Android with Kotlin

### LINKED ISSUE #3440: Geofencing In Android with Kotlin
### Introduction paragraph (2-3 paragraphs):
Geofences are a powerful tool in a developer’s  Android development process. Geofences give devices the power to monitor a circular area in the world, and let the device inform you whenever it enters or exits that area.

Geofencing combines awareness of the user’s current location with awareness of nearby features, defined as the user’s proximity to locations that may be of interest. To mark a location of interest, you specify its latitude and longitude. To adjust the proximity for the location, you add a radius. The latitude, longitude, and radius define a geofence. You can have multiple active geofences at one time.
In this tutorial we will focus on setting up geofences in android application.

### Key takeaways: 
- What is geofencing
- Creating custom geofences
- Setting up geofencing in android
- Integrating notification with geofencing

### References: 
- [Android documentation](https://developer.android.com/training/location/geofencing)

### PR DESCRIPTION
**Pull request by brandy-kay**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x]  Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #3440

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by brandy-kay** [comment]
hello @Ericgacoki have added hero image

**Comment by ahmadmardeni1** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Geofencing in Android with Kotlin](https://github.com/section-engineering-education/engineering-education/files/7169607/New.Scan.11_17.AM.Copyleaks.Report.pdf)

**Comment by brandy-kay** [comment]
hello @ahmadmardeni1 and @Ericgacoki i have fixed the plagiarism the most percentage is raised by code part which can't be changed because they are  built-in methods.

**Comment by ahmadmardeni1** [comment]
The % is still high, but most of it is the code. Please let me know if it can't be changed/edited as Brandy said. @Ericgacoki 

[Geofencing in Android with Kotlin](https://github.com/section-engineering-education/engineering-education/files/7189708/New.Scan.10_05.AM.Copyleaks.Report.pdf)

**Comment by Ericgacoki** [comment]
@ahmadmardeni1,
As Brandy mentioned, most of the code consist of inbuilt and lifecycle callback methods that can't be edited. These methods are applicable in numerous use cases including Geo-Fencing. I think that's what is causing a high plagiarism.

Do you think I should proceed with the review?

**Comment by ahmadmardeni1** [comment]
Okay, then no problem. 
Yes, proceed with the review, please. @Ericgacoki

File: content/articles/geofencing-in-android-with-kotlin/index.md [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+```kotlin
+ private fun askLocationPermission() {
+        if (authorizedLocation())
+            return
+        var grantingPermission = arrayOf(Manifest.permission.ACCESS_FINE_LOCATION)
+        val customResult = when {
+            gadgetQ -> {
+                grantingPermission += Manifest.permission.ACCESS_BACKGROUND_LOCATION
+                REQUEST_FOREGROUND_AND_BACKGROUND_PERMISSION_RESULT_CODE
+            }
```
Comment (Ericgacoki) [inline content/articles/geofencing-in-android-with-kotlin/index.md:107]: I can't seem to find where these codes (Request code and Result code) were initialized 🤭. Kindly assign them some values.

File: content/articles/geofencing-in-android-with-kotlin/index.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+```kotlin
+private fun validateGadgetAreaInitiateGeofence(resolve: Boolean = true) {
+        val locationRequest = LocationRequest.create().apply {
+            priority = LocationRequest.PRIORITY_LOW_POWER
+        }
+        val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
+
+        val client = LocationServices.getSettingsClient(this)
+        val locationResponses =
+            client.checkLocationSettings(builder.build())
```
Comment (Ericgacoki) [inline content/articles/geofencing-in-android-with-kotlin/index.md:149]: ...
I think a lot is happening in this snippet. Add an explanation either via comments or a plain text just below the code.

File: content/articles/geofencing-in-android-with-kotlin/index.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+geoClient = LocationServices.getGeofencingClient(this)
+```
+
+Also within the onCreate method, add a geofenceList that holds geofences. In this step, we will add one geofence but you can have many geofences. 
+
+```kotlin
+geofenceList.add(Geofence.Builder()
+            .setRequestId("entry.key")
+            .setCircularRegion(latitude,longitude,radius)
+            .setExpirationDuration(Geofence.NEVER_EXPIRE)
```
Comment (Ericgacoki) [inline content/articles/geofencing-in-android-with-kotlin/index.md:215]: These arguments are also not initialized. Kindly give a sample value for the `latitude`, `longitude`, and `radius`.

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
Hi @brandy-kay, kindly fix the following issues and ping me when you're done!

**Comment by brandy-kay** [comment]
Hello @Ericgacoki  have made the requested changes

**Review by Ericgacoki — APPROVED** [review]

**Comment by WanjaMIKE** [comment]
Merged #3819 into master

**Pull request merged**